rskj-core/src/main/java/co/rsk/peg/federation/FederationStorageProviderImpl.java

```diff
@@ -0,0 +1,374 @@
+        return federationElection;
+    }
+    @Override
+    public Optional<Long> getActiveFederationCreationBlockHeight(ActivationConfig.ForBlock activations) {
+        if (!activations.isActive(RSKIP186)) {
+            return Optional.empty();
+        }
+
+        if (activeFederationCreationBlockHeight != null) {
+            return Optional.of(activeFederationCreationBlockHeight);
+        }
+
+        activeFederationCreationBlockHeight = bridgeStorageAccessor.safeGetFromRepository(ACTIVE_FEDERATION_CREATION_BLOCK_HEIGHT_KEY.getKey(), BridgeSerializationUtils::deserializeOptionalLong).orElse(null);
+        return Optional.ofNullable(activeFederationCreationBlockHeight);
+    }
+    @Override
+    public void setActiveFederationCreationBlockHeight(long activeFederationCreationBlockHeight) {
+        this.activeFederationCreationBlockHeight = activeFederationCreationBlockHeight;
+    }
+
+    @Override
+    public Optional<Long> getNextFederationCreationBlockHeight(ActivationConfig.ForBlock activations) {
+        if (!activations.isActive(RSKIP186)) {
+            return Optional.empty();
+        }
+
+        if (nextFederationCreationBlockHeight != null) {
+            return Optional.of(nextFederationCreationBlockHeight);
+        }
+
+        nextFederationCreationBlockHeight = bridgeStorageAccessor.safeGetFromRepository(NEXT_FEDERATION_CREATION_BLOCK_HEIGHT_KEY.getKey(), BridgeSerializationUtils::deserializeOptionalLong).orElse(null);
+        return Optional.ofNullable(nextFederationCreationBlockHeight);
+    }
+    @Override
+    public void setNextFederationCreationBlockHeight(long nextFederationCreationBlockHeight) {
+        this.nextFederationCreationBlockHeight = nextFederationCreationBlockHeight;
+    }
+    public void clearNextFederationCreationBlockHeight() {
+        this.nextFederationCreationBlockHeight = -1L;
+    }
+
+    @Override
+    public Optional<Script> getLastRetiredFederationP2SHScript(ActivationConfig.ForBlock activations) {
+        if (!activations.isActive(RSKIP186)) {
+            return Optional.empty();
+        }
+
+        if (lastRetiredFederationP2SHScript != null) {
+            return Optional.of(lastRetiredFederationP2SHScript);
+        }
+
+        lastRetiredFederationP2SHScript = bridgeStorageAccessor.safeGetFromRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), BridgeSerializationUtils::deserializeScript);
+        return Optional.ofNullable(lastRetiredFederationP2SHScript);
+    }
+    @Override
+    public void setLastRetiredFederationP2SHScript(Script lastRetiredFederationP2SHScript) {
+        this.lastRetiredFederationP2SHScript = lastRetiredFederationP2SHScript;
+    }
+
+    @Override
+    public void save(NetworkParameters networkParameters, ActivationConfig.ForBlock activations) {
```

Comment:
Probably the activations object should simply be part of the class. The save method will need to be called eventually, so in any case it required to have an activations object in order to interact with this class

Reply:
yes, but this `save()` method will just be called from `FederationSupport`, that already has `activations` as part of it

Reply:
Ok, makes sense. We only request the activations object when it's required. Thanks